File: src/main/java/de/axelrindle/chickendropfeathersmod/EntityAIDropFeather.java
package de.axelrindle.chickendropfeathersmod;

import net.minecraft.entity.ai.EntityAIBase;
import net.minecraft.entity.passive.EntityChicken;
import net.minecraft.init.Items;
import net.minecraft.item.ItemStack;

import java.util.concurrent.ThreadLocalRandom;

public class EntityAIDropFeather extends EntityAIBase {

    private final ThreadLocalRandom random = ThreadLocalRandom.current();

    private final int baseTicks = 12000;
    private final EntityChicken chicken;
    private int timeUntilNextFeather = getNewTime();

    public EntityAIDropFeather(EntityChicken chicken) {
        this.chicken = chicken;
    }

    @Override
    public boolean shouldExecute() {
        return !chicken.isChild() && !chicken.isChickenJockey();
    }

    @Override
    public void resetTask() {
        timeUntilNextFeather = getNewTime();
    }

    @Override
    public void tick() {
        timeUntilNextFeather--;
        System.out.println("Feathers in " + timeUntilNextFeather + " ticks...");
        if (timeUntilNextFeather <= 0) {
            dropFeather();
            resetTask();
        }
    }

    /**
     * Returns a new tick delay until the next feather will be dropped.
     *
     * @return An integer between 12000 and 24000
     * @see #baseTicks
     */
    private int getNewTime() {
        return random.nextInt(baseTicks, baseTicks * 2 + 1);
    }

    /**
     * Returns the amount of feathers that should be dropped.
     *
     * @return An integer between 1 and 2.
     */
    private int getFeatherAmount() {
        return random.nextInt(3);
    }

    private void dropFeather() {
        chicken.entityDropItem(new ItemStack(Items.FEATHER, getFeatherAmount()));
    }
}

File: src/main/kotlin/de/axelrindle/chickendropfeathersmod/EntityAIDropFeather.kt
package de.axelrindle.chickendropfeathersmod

import net.minecraft.entity.ai.EntityAIBase
import net.minecraft.entity.passive.EntityChicken
import net.minecraft.init.Items
import java.util.*

/**
 * A custom [EntityAIBase] instance responsible for dropping a feather from
 * time to time.
 */
class EntityAIDropFeather(

        /**
         * The [EntityChicken] this AI belongs to.
         */
        private val chicken: EntityChicken

) : EntityAIBase() {

    private val rand: Random = Random()

    /**
     * The base amount of ticks used for time calculation.
     *
     * 20 ticks equal 1 second. 12000 ticks equal 600 seconds. 600 Seconds equal 10 minutes.
     */
    private val baseTicks: Int = 1000
    private var timeUntilNextFeather: Int = getNewTime()


    override fun shouldExecute(): Boolean {
        return true
    }

    override fun resetTask() {
        timeUntilNextFeather = getNewTime()
    }

    override fun updateTask() {
        timeUntilNextFeather--

        if (timeUntilNextFeather == 0) {
            dropFeather()
            resetTask()
        }
    }

    /**
     * Returns a new tick delay until the next feather will be dropped.
     * Currently this lies between 10 and 20 minutes.
     *
     * @return An integer between 12000 and 23999
     * @see baseTicks
     */
    private fun getNewTime(): Int {
        return this.rand.nextInt(baseTicks) + baseTicks
    }

    /**
     * Returns the amount of feathers that should be dropped.
     *
     * @return An integer between 1 and 2.
     */
    private fun getFeatherAmount(): Int {
        return this.rand.nextInt(2) + 1
    }

    private fun dropFeather() {
        chicken.dropItem(Items.FEATHER, getFeatherAmount())
    }
}